akka-actor/src/main/scala/akka/util/MessageBuffer.scala

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.util
+
+import akka.actor.ActorRef
+
+/**
+ * A single producer single consumer mutable message buffer
+ */
+final class MessageBuffer private (
+  private var _head: MessageBuffer.Node,
+  private var _tail: MessageBuffer.Node) {
```

Comment:
It can be more efficient to loop array based structures. Isn't the typical usage pattern to append elements and finally `foreach` all of them followed by `clear`.
That can be allocation free if backed by two `ArrayDeque`.

For very large buffers the linked list may still be better.